### PR TITLE
Refactor triangulationWrapper

### PR DESCRIPTION
**Pull request by lucacarniato**

### CODE REVIEW

File: libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp [hunk 1]
```diff
@@ -83,117 +73,165 @@ namespace meshkernel
...
+        }
 
-            // Create edges
-            if (m_numEdges == 0)
-            {
-                return;
-            }
+        /// @brief Gets the number of triangulated faces
+        /// @return The number of triangulated faces
+        [[nodiscard]] const auto& GetNumFaces() const
```
Comment (ahmad-el-sayed) [inline libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp:147]: Return by value, it's a just an int.

File: libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp [hunk 2]
```diff
@@ -83,117 +73,165 @@ namespace meshkernel
...
-                m_faceNodes[f][1] = static_cast<size_t>(faceNodesFlat[faceCounter] - 1);
-                m_faceEdges[f][1] = static_cast<size_t>(faceEdgesFlat[faceCounter] - 1);
-                faceCounter++;
-                m_faceNodes[f][2] = static_cast<size_t>(faceNodesFlat[faceCounter] - 1);
-                m_faceEdges[f][2] = static_cast<size_t>(faceEdgesFlat[faceCounter] - 1);
-                faceCounter++;
-            }
+        /// @brief Gets the number of triangulated nodes
+        /// @return The number of triangulated nodes
+        [[nodiscard]] const auto& GetNumNodes() const
```
Comment (ahmad-el-sayed) [inline libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp:140]: Return by value, it's a just an int.

File: libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp [hunk 3]
```diff
@@ -83,117 +73,165 @@ namespace meshkernel
...
 
-            // Create nodes
-            m_nodes.resize(m_numNodes);
-            for (size_t i = 0; i < m_numNodes; ++i)
-            {
-                m_nodes[i] = {xNodesFlat[i], yNodesFlat[i]};
-            }
+        /// @brief Gets the number of triangulated edges
+        /// @return The number of triangulated edges
+        [[nodiscard]] const auto& GetNumEdges() const
```
Comment (ahmad-el-sayed) [inline libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp:133]: Return by value, it's a just an int.
Personally, I'm not big fan of return type deduction. It makes code harder to read and maintain.

File: libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp [hunk 4]
```diff
@@ -83,117 +73,165 @@ namespace meshkernel
...
-                    {
-                        m_edgesFaces[edge][1] = f;
-                    }
-                }
-            }
+        /// @brief Retrieves the face node
+        /// @param faceIndex The index of the face to retrieve the node from
+        /// @param nodeIndex The index of the node to retrieve
+        /// @return const reference to the node with the specified index for the specified face
+        [[nodiscard]] const auto& GetFaceNode(const size_t& faceIndex, const size_t& nodeIndex) const
```
Comment (ahmad-el-sayed) [inline libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp:171]: Can drop the constant reference in the return type, it's a size_t.
Pass indices by value, no need for constant reference either.

File: libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp [hunk 5]
```diff
@@ -83,117 +73,165 @@ namespace meshkernel
...
+        [[nodiscard]] const auto& GetEdgeNode(const size_t& edgeIndex, const size_t& nodeIndex) const
+        {
+            return m_edgeNodes[edgeIndex][nodeIndex];
+        }
+
+        /// @brief Retrieves the edge face
+        /// @param edgeIndex The index of the edge to retrieve the node from
+        /// @param faceIndex The index of the face to retrieve
+        /// @return const reference to the edge with the specified index for the specified face
+        [[nodiscard]] const auto& GetEdgeFace(const size_t& edgeIndex, const size_t& faceIndex) const
```
Comment (ahmad-el-sayed) [inline libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp:198]: same as previous comment

File: libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp [hunk 6]
```diff
@@ -83,117 +73,165 @@ namespace meshkernel
...
+        {
+            return m_edgesFaces[edgeIndex][faceIndex];
+        }
+
+        /// @brief Retrieves the x coordinate of a triangulated node
+        /// @param nodeIndex The index of the node to retrieve
+        /// @return const reference to the x coordinate
+        [[nodiscard]] const auto& GetXCoord(const size_t& nodeIndex) const
```
Comment (ahmad-el-sayed) [inline libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp:206]: same as previous comment

File: libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp [hunk 7]
```diff
@@ -83,117 +73,165 @@ namespace meshkernel
...
+        std::vector<int> m_edgeNodesFlat; ///< Edge nodes flat array passed to the triangulation library
+        std::vector<int> m_faceEdgesFlat; ///< Face edges flat array passed to the triangulation library
+        std::vector<double> m_xCoordFlat; ///< x coordinates flat array passed to the triangulation library
+        std::vector<double> m_yCoordFlat; ///< y coordinates flat array passed to the triangulation library
+        int m_numNodes{0};                ///< Initial number of triangulated nodes
+        int m_numEdges{0};                ///< Initial number of triangulated edges
+        int m_numFaces{0};                ///< Initial number of triangulated faces
+
+        std::vector<Point> m_nodes;                    ///< Reconstructed vector of nodes
+        std::vector<std::vector<size_t>> m_faceNodes;  ///< Reconstructed vector of face nodes
```
Comment (ahmad-el-sayed) [inline libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp:231]: Nothing to to do, just a note.
I've seen vector of vectors used a lot throughout the code. We can:
1. define out own matrix class Matrix (contiguous memory) with any necessary operators.
2. use boost::numeric::ublas::matrix
3. use the eigen library

File: libs/MeshKernel/tests/src/Mesh2DTest.cpp [hunk 8]
```diff
@@ -77,47 +77,6 @@ TEST(Mesh2D, OneQuadTestConstructor)
...
-    ASSERT_EQ(2, mesh.m_edges[3].second);
-
-    ASSERT_EQ(2, mesh.m_edges[4].first);
-    ASSERT_EQ(3, mesh.m_edges[4].second);
-
-    ASSERT_EQ(3, mesh.m_edges[5].first);
-    ASSERT_EQ(0, mesh.m_edges[5].second);
-}
-
 TEST(Mesh2D, TriangulateSamples)
```
Comment (ahmad-el-sayed) [inline libs/MeshKernel/tests/src/Mesh2DTest.cpp:80]: This unit test does not seem to test anything.

File: libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp (hunk 7 above)
Comment (lucacarniato) [reply libs/MeshKernel/include/MeshKernel/TriangulationWrapper.hpp:233]: indeed, i agree

File: libs/MeshKernel/tests/src/Mesh2DTest.cpp (hunk 8 above)
Comment (lucacarniato) [reply libs/MeshKernel/tests/src/Mesh2DTest.cpp:80]: it is checking the correct triangulation is produced

**Review by ahmad-el-sayed — APPROVED** [review]

**Pull request merged**